### ISSUE TITLE
support for copying custom values from `OAuth2TokenContext` to `JwtEncodingContext`

### ISSUE THREAD

**Issue by emilburzo** [spring-projects/spring-authorization-server#1154]
**Expected Behavior**
In [spring-attic/spring-security-oauth](https://github.com/spring-attic/spring-security-oauth), specifically in `TokenEnhancer`, it was possible to add claims to the JWT token based on the refresh token value, as it was accessible via `OAuth2AccessToken.getRefreshToken()`

I was expecting it would be possible via `OAuth2TokenContext.getAuthorization()` but this doesn't seem to be set (always null) for the grants I've tried.

**Current Behavior**
It doesn't seem like there's any way to access the refresh token value from `OAuth2TokenCustomizer<JwtEncodingContext>`.

Another option I've tried is to add a custom key/value in `OAuth2TokenContext` by using `DefaultOAuth2TokenContext.Builder.put(key, value)`, but it seems like custom values are not copied over to the `JwtEncodingContext` in [JwtGenerator.generate()](https://github.com/spring-projects/spring-authorization-server/blob/8c63c7dbafc6db764532aa2527b959ef4f2145f7/oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/token/JwtGenerator.java#L75), only known ones. 

Would it be possible to add support for something like `additionalParameters` to handle copying custom values from `OAuth2TokenContext` to `JwtEncodingContext`?

Otherwise it seems like the only option would be to have a custom `JwtGenerator` and use that instead, or did I miss an easier approach?

**Context**
I'm migrating a project using the existing library, which added a computed claim to the JWT based on the refresh token value.

**Comment by jgrandja** [comment]
@emilburzo 

> it was possible to add claims to the JWT token based on the refresh token value, as it was accessible via `OAuth2AccessToken.getRefreshToken()`

The default `OAuth2RefreshTokenGenerator` generates an opaque token so I don't think it will be useful in this case.

> I was expecting it would be possible via OAuth2TokenContext.getAuthorization() but this doesn't seem to be set (always null) for the grants I've tried.

The `OAuth2RefreshToken` is generated after the `OAuth2AccessToken`.

> Would it be possible to add support for something like `additionalParameters` to handle copying custom values from `OAuth2TokenContext` to `JwtEncodingContext`?

Before we consider an enhancement, I'd like to see if we can make things work with the existing customization hooks. Can you please provide more details on your legacy implementation? What are you looking up in the refresh token and using/copying to the access token?

**Comment by emilburzo** [comment]
@jgrandja 

> Before we consider an enhancement, I'd like to see if we can make things work with the existing customization hooks.

FWIW, I was able to proceed with a custom `JwtGenerator`. You are also correct that I had to generate the refresh token before the access token, but since so far I've only needed it for the unsupported `password` grant, which must be self-implemented anyway, that wasn't a blocker.

There's code duplication, but now it's more of a question of convenience/maintainability than a necessity.

> What are you looking up in the refresh token and using/copying to the access token?

The legacy implementation added a custom identifier based on the refresh token so that it can be uniquely identified from "the outside".

Use-case: users of the refresh token have the ability to request expiry of all refresh tokens except one (e.g. "logout" everywhere except one session)

**Comment by jgrandja** [comment]
@emilburzo 

> The legacy implementation added a custom identifier based on the refresh token so that it can be uniquely identified from "the outside".

I'm assuming the legacy implementation generated a JWT refresh token since it needed to carry a custom identifier as a claim? I'm also assuming you made this work using a custom `OAuth2TokenGenerator` instead of using the default `OAuth2RefreshTokenGenerator`:

> FWIW, I was able to proceed with a custom `JwtGenerator`

If my understanding is correct, then the customization you applied is the correct path. Given that `OAuth2RefreshTokenGenerator` supports opaque token values only, you would need to implement your own `OAuth2TokenGenerator` to generate a self-contained token value for refresh tokens. Also, there are no plans to provide a `OAuth2TokenGenerator` implementation for JWT refresh tokens.

It sounds like we can close this issue if my understanding is correct since JWT refresh tokens would need to be a customization applied in the application. I'll leave it open for now until you respond.

**Comment by emilburzo** [comment]
@jgrandja 

> I'm assuming the legacy implementation generated a JWT refresh token since it needed to carry a custom identifier as a claim? I'm also assuming you made this work using a custom OAuth2TokenGenerator instead of using the default OAuth2RefreshTokenGenerator:

sorry, it seems I explained the legacy implementation poorly, let me rephrase:

- the legacy implementation generated an opaque **refresh token**
- based on the **refresh token** value, a unique identifier was added as a claim to the **access token** JWT using some sort of hashing scheme

(I cannot provide any insights on why it was designed this way, as that knowledge is gone)

The enable for this in the legacy implementation is that the `TokenEnhancer` had access to the refresh token value, whereas the new `JwtGenerator` doesn't because, as you mentioned, the refresh token is generated after the access token

**Comment by jgrandja** [comment]
Thanks for the explanation @emilburzo. I now understand your issue.

I'm reluctant on modifying the existing implementations `OAuth2AuthorizationCodeAuthenticationProvider` and  `OAuth2RefreshTokenAuthenticationProvider` to change the ordering of token generation - it doesn't make sense to generate the refresh token before the access token.

But it sounds like you're not looking for these changes since this is related to the `password` grant?

> FWIW, I was able to proceed with a custom `JwtGenerator`. You are also correct that I had to generate the refresh token before the access token, but since so far I've only needed it for the unsupported `password` grant, which must be self-implemented anyway, that wasn't a blocker.

If your case is directly related to a `password` grant implementation, then you have complete control in the call order of the `OAuth2TokenGenerator` so you can generate the refresh token first and then associate it to `OAuth2Authorization`, which then would be available when the `OAuth2TokenGenerator` is subsequently called for access token generation - you just need to ensure the `OAuth2TokenContext.OAuth2Authorization` includes the refresh token.

Is this how you implemented it?

Take a look at `OAuth2AuthorizationCodeAuthenticationProvider` as it has a similar use case where the ID Token generation MAY need access to the access and/or refresh token:

https://github.com/spring-projects/spring-authorization-server/blob/098ab85e2a830966d7ac343d606c1b9d062eb54a/oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java#L229

**Comment by emilburzo** [comment]
@jgrandja 

> I'm reluctant on modifying the existing implementations OAuth2AuthorizationCodeAuthenticationProvider and OAuth2RefreshTokenAuthenticationProvider to change the ordering of token generation - it doesn't make sense to generate the refresh token before the access token.

Understood, and since apparently nobody else ran into this problem, it's what I expected.

> But it sounds like you're not looking for these changes since this is related to the password grant?

Correct.

> If your case is directly related to a password grant implementation, then you have complete control in the call order of the OAuth2TokenGenerator so you can generate the refresh token first and then associate it to OAuth2Authorization, which then would be available when the OAuth2TokenGenerator is subsequently called for access token generation - you just need to ensure the OAuth2TokenContext.OAuth2Authorization includes the refresh token.

Nice, this is actually easier than what I did :point_down:. I had a suspicion this approach might be a problem for the `refresh_token` grant, since I discovered it's also needed there, but the `authorization` is already set in the `refresh_token` grant, so I have access to the refresh token in the JWT customizer -- all good :+1:

> Is this how you implemented it?

It looks like I went the harder way, I created a custom `OAuth2TokenGenerator<Jwt>` with the only change being in the `generate()` function, to copy the custom id from `OAuth2TokenContext` to the `JwtEncodingContext`.

But with your cleaner solution from above, I can delete this part and this github issue can be closed from my side.

Thank you for all the support!

**Comment by jgrandja** [comment]
@emilburzo I'm glad I was able to help you find an easier solution for your custom implementation 👍 

I'll close this issue.
